### PR TITLE
[WIP] [coverage] Switch to llvm-profdata's pooled profdata merge

### PR DESCRIPTION
**Pull request by harlanhaskins**
#### What's in this pull request?

This removes the now-obsolete `profdata_merge` server, instead using LLVM's built-in profile data pooling. This accomplishes a few things:
- LLVM's pooling is very efficient, so there's very little overhead when running the test suite
- We don't have to maintain that `profdata_merge` server anymore.
- LLVM's pooling is natively supported on Linux, and works right now with this patch.

The one downside is that you'll need to use a recent upstream build of clang/llvm to build and test coverage, which we can ensure exist on the bots. I still need to add a guard in build-script to prevent people from trying to run this with an old version of llvm-profdata. 
#### Resolved bug number: (N/A)

---

Before merging this pull request to apple/swift repository:
- [ ] Test pull request on Swift continuous integration.

<details>
  <summary>

Triggering Swift CI</summary>

The swift-ci is triggered by writing a comment on this PR addressed to the GitHub user @swift-ci. Different tests will run depending on the specific comment that you use. The currently available comments are:

**Smoke Testing**

| Platform | Comment |
| --- | --- |
| All supported platforms | @swift-ci Please smoke test |
| All supported platforms | @swift-ci Please smoke test and merge |
| OS X platform | @swift-ci Please smoke test OS X platform |
| Linux platform | @swift-ci Please smoke test Linux platform |

**Validation Testing**

| Platform | Comment |
| --- | --- |
| All supported platforms | @swift-ci Please test |
| All supported platforms | @swift-ci Please test and merge |
| OS X platform | @swift-ci Please test OS X platform |
| OS X platform | @swift-ci Please benchmark |
| Linux platform | @swift-ci Please test Linux platform |

**Lint Testing**

| Language | Comment |
| --- | --- |
| Python | @swift-ci Please Python lint |

Note: Only members of the Apple organization can trigger swift-ci.
</details>

### CODE REVIEW

**Comment by harlanhaskins** [comment]
@gribozavr @vedantk Thoughts?

File: test/CMakeLists.txt [hunk 1]
```diff
@@ -147,25 +147,12 @@ if(PYTHONINTERP_FOUND)
       set(VARIANT_SDK "${SWIFT_SDK_${SDK}_PATH}")
 
       # A directory where to put the xUnit-style XML test results.
-      set(swift_test_results_dir
+      set(SWIFT_TEST_RESULTS_DIR
```
Comment (gribozavr) [inline test/CMakeLists.txt]: It was lowercase because it is a "local".

**Comment by gribozavr** [comment]
LGTM.

File: test/CMakeLists.txt (hunk 1 above)
Comment (harlanhaskins) [inline test/CMakeLists.txt]: It's no longer a local; I need to get it inside `lit`. Though, if there's a better way to get to that directory, I'd be happy to use it!

Comment (gribozavr) [inline test/CMakeLists.txt]: Ah, OK!

File: test/lit.site.cfg.in [hunk 2]
```diff
@@ -29,6 +29,7 @@ config.variant_triple = "@VARIANT_TRIPLE@"
 config.variant_sdk = "@VARIANT_SDK@"
 config.variant_suffix = "@VARIANT_SUFFIX@"
 config.swiftlib_dir = "@LIT_SWIFTLIB_DIR@"
+config.swift_test_results_dir = "@SWIFT_TEST_RESULTS_DIR@"
```
Comment (rintaro) [inline test/lit.site.cfg.in]: Could you make this `--param` configurable?
```
lit_config.params.get("swift_test_results_dir", "@SWIFT_TEST_RESULTS_DIR@")
```

Comment (rintaro) [inline test/lit.site.cfg.in]: Never mind. I will do that. I must modify `utils/run-test` anyway.

Comment (harlanhaskins) [inline test/lit.site.cfg.in]: I'll merge this now, then.

**Comment by harlanhaskins** [comment]
@swift-ci please test and merge

File: test/lit.cfg [hunk 3]
```diff
@@ -153,24 +149,23 @@ class SwiftTest(lit.formats.ShTest, object):
...
-            test.config.environment["LLVM_PROFILE_FILE"] = \
-                os.path.join(profdir, "swift-%p.profraw")
+            _, tmp_base = lit.TestRunner.getTempPaths(test)
+            if self.coverage_mode == "NOT_MERGED":
+                profdir = tmp_base + '.profdir'
+                if not os.path.exists(profdir):
+                    os.makedirs(profdir)
+
+                test.config.environment["LLVM_PROFILE_FILE"] = \
+                    os.path.join(profdir, "swift-%p.profraw")
```
Comment (vedantk) [inline test/lit.cfg]: Watch out for PID wraparound. This bit Matthias when he was working on lnt -- it's not just a theoretical limit.

Comment (harlanhaskins) [inline test/lit.cfg]: Would a conflict arise if we named the test, say, `<test-source-file>-%p.profraw`? That would make it easier to identify which prof data corresponded to which test.

**Comment by harlanhaskins** [comment]
@swift-ci please test

File: test/lit.site.cfg.in (hunk 2 above)
Comment (harlanhaskins) [inline test/lit.site.cfg.in]: I had to update the diff anyway, so I went ahead and made this change.

**Pull request merged**